### DISCUSSION TITLE
patches_in_band

### DISCUSSION THREAD

**Discussion by Newshamod** [UMEP-dev/UMEP#435] [General]
**Hi there, 

I have tried to run SOLWEIG model but I am receiving the following error:**

Algorithm started at: 2022-07-07T14:53:38
Algorithm 'Outdoor Thermal Comfort: SOLWEIG' starting…
Input parameters:
{ 'ABS_L' : 0.95, 'ABS_S' : 0.7, 'ACTIVITY' : 80, 'AGE' : 35, 'ALBEDO_GROUND' : 0.15, 'ALBEDO_WALLS' : 0.2, 'CLO' : 0.9, 'CYL' : True, 'EMIS_GROUND' : 0.95, 'EMIS_WALLS' : 0.9, 'HEIGHT' : 180, 'INPUTMET' : 'D:\\Newsha\\LiDAR\\for qgis\\grid1\\met_solweig.txt', 'INPUT_ANISO' : '', 'INPUT_ASPECT' : 'D:/Newsha/LiDAR/for qgis/grid1/WallA_grid1.tif', 'INPUT_CDSM' : 'D:/Newsha/LiDAR/lidar_dataset/CDSM_grid1.tif', 'INPUT_DEM' : 'D:/Newsha/LiDAR/lidar_dataset/DEM_grid1.tif', 'INPUT_DSM' : 'D:/Newsha/LiDAR/lidar_dataset/DSMb_grid1.tif', 'INPUT_HEIGHT' : 'D:/Newsha/LiDAR/for qgis/grid1/WallH_grid1.tif', 'INPUT_LC' : None, 'INPUT_SVF' : 'D:\\Newsha\\LiDAR\\for qgis\\grid1\\SVF_g1\\svfs.zip', 'INPUT_TDSM' : None, 'INPUT_THEIGHT' : 25, 'ONLYGLOBAL' : False, 'OUTPUT_DIR' : 'D:\\Newsha\\LiDAR\\for qgis\\grid1\\MRT_g1', 'OUTPUT_KDOWN' : True, 'OUTPUT_KUP' : True, 'OUTPUT_LDOWN' : True, 'OUTPUT_LUP' : True, 'OUTPUT_SH' : True, 'OUTPUT_TMRT' : True, 'OUTPUT_TREEPLANTER' : False, 'POI_FIELD' : '', 'POI_FILE' : None, 'POSTURE' : 0, 'SAVE_BUILD' : False, 'SENSOR_HEIGHT' : 10, 'SEX' : 0, 'TRANS_VEG' : 3, 'USE_LC_BUILD' : False, 'UTC' : 1, 'WEIGHT' : 75 }

Longitude derived from DSM: -4.392889406995406
Latitude derived from DSM: 55.77399297898338
Vegetation scheme activated
Sky View Factor rasters loaded
Meteorological data successfully loaded
Calculating sun positions for each time step
Isotropic sky
Writing settings for this model run to specified output folder (Filename: RunInfoSOLWEIG_YYYY_DOY_HHMM.txt)
Executing main model
Traceback (most recent call last):
File "C:\Users/nmodjr200/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\processing_umep\processor\solweig_algorithm.py", line 955, in processAlgorithm
dRad, Kside = so.Solweig_2022a_calc(
File "C:\Users/nmodjr200/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\processing_umep\functions\SOLWEIGpython\Solweig_2022a_calc_forprocessing.py", line 133, in Solweig_2022a_calc
lv, pc_, pb_ = Perez_v3(zenDeg, azimuth, radD, radI, jday, patchchoice, patch_option) # Relative luminance
File "C:\Users/nmodjr200/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\processing_umep\util\SEBESOLWEIGCommonFiles\Perez_v3.py", line 193, in Perez_v3
skyvaultalt, skyvaultazi, _, _, _, _, _ = create_patches(patch_option)
File "C:\Users/nmodjr200/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\processing_umep\util\SEBESOLWEIGCommonFiles\create_patches.py", line 41, in create_patches
skyvaultaziint = np.array([360/patches for patches in patches_in_band])
UnboundLocalError: local variable 'patches_in_band' referenced before assignment

Execution failed after 4.41 seconds

Loading resulting layers
Algorithm 'Outdoor Thermal Comfort: SOLWEIG' finished

**Could you please help me with this issue? and here is the output in the directory folder:**

This file provides run settings for the SOLWEIG run initiated at: Thu, 07 Jul 2022 14:53:42
Version: SOLWEIG v2021a

SURFACE DATA
Digital surface model (DSM): D:/Newsha/LiDAR/lidar_dataset/DSMb_grid1.tif
Model domain: rows = 3000, columns = 3501
Projected reference system: British_National_Grid
Geographical coordinate system: OSGB36
Latitude: 55.77399297898338
Longitude: -4.392889406995406
UTC: 1.0
Transmissivity of light through vegetation: 0.03
Digital vegetation canopy model (CDSM): D:/Newsha/LiDAR/lidar_dataset/CDSM_grid1.tif
Trunkzone estimated from CDSM
Trunkzone as percent of canopy height: 0.25
Landcover scheme inactive

DEM used to identify buildings

METEOROLOGICAL FORCING DATA
Meteorological file: D:\Newsha\LiDAR\for qgis\grid1\met_solweig.txt

HUMAN EXPOSURE PARAMETERS
Absorption, shortwave radiation: 0.7
Absorption, longwave radiation: 0.95
Posture of human body: Standing
ENVIRONMENTAL PARAMETERS
Albedo of walls: 0.2
Albedo of ground (not used if land cover scheme is active): 0.15
Emissivity (walls): 0.9
Emissivity of ground (not used if land cover scheme is active): 0.95

ADDITIONAL SETTINGS
Human considered as a standing cylinder
Isotropic sky

**Thanks**

**Comment by biglimp** [comment]
Try the following:
1. Update UMEP. There is new solweig version out (v2022a).
2. Set correct UTC. I guess you are in UTC=0 (Glasgow)
If the above does not work, Disable the anisotrophic scheme.

**Reply by Newshamod** [reply]
The UMEP which I'm working on, is the latest version and even with UTC of 0, I received the same error.

Could you please tell me how I can disable the anisotrophic scheme? Is it related to the continuous meteorological dataset as I set the data for 24 hour? 

[met_solweig.txt](https://github.com/UMEP-dev/UMEP/files/9066331/met_solweig.txt)

**Reply by biglimp** [reply]
Yes, your metfile is not correct. The it (hour) and imin (minute) looks strange. If you are supposed to have hourly resolution, it column should increment by 1 and imin column should be 0 at all positions. See https://umep-docs.readthedocs.io/en/latest/pre-processor/Meteorological%20Data%20MetPreprocessor.html for more info on metfiles.

**Reply by Newshamod** [reply]
Oh my bad....
I followed the instruction and cleaned the metfile but got the following error:

 EXCEPTION IN C:\Users/nmodjr200/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\UMEP\SOLWEIG\solweigworker.py, 
             LINE 290 "dRad, Kside = so.Solweig_2022a_calc(" 
             ERROR MESSAGE: local variable 'L_patches' referenced before assignment

**Reply by biglimp** [reply]
Then I can think of no other reason that you have the wrong version. I see from your previous post that you are using 2021a and not 2022a. Update to UMEP 3.20 which includes solweig v2022a. If your error presists, please submit an issue including data and explanarions.

**Reply by Newshamod** [reply]
Thank you for your comments. Unfortunately, with the latest version, I received the same error. I am posting it now as an issue with all the data.

**Reply by biglimp** [reply]
Ok, @nilswallenberg who has developed this part will look into this when he is back from vacation..

**Reply by Newshamod** [reply]
I appreciate your support and time @biglimp..

**Comment by Axou973** [comment]
Hello, I have the same problem. With the UMEP processor from Processing ToolBox, The outdoor thermal comfort doesn't work.
they show me that : 
Longitude derived from DSM: 4.78884444108657
Latitude derived from DSM: 45.77444574049266
Vegetation scheme activated
Sky View Factor rasters loaded
Meteorological data successfully loaded
Calculating sun positions for each time step
Isotropic sky
Writing settings for this model run to specified output folder (Filename: RunInfoSOLWEIG_YYYY_DOY_HHMM.txt)
Executing main model
Traceback (most recent call last):
**File "C:\Users/admin-ddavid/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\processing_umep\processor\solweig_algorithm.py", line 955, in processAlgorithm
dRad, Kside = so.Solweig_2022a_calc(
File "C:\Users/admin-ddavid/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\processing_umep\functions\SOLWEIGpython\Solweig_2022a_calc_forprocessing.py", line 344, in Solweig_2022a_calc
Lside, L_patches, CI_Tg, CI_TgG, KsideD, dRad, Kside
UnboundLocalError: local variable 'L_patches' referenced before assignment

Execution failed after 6.06 secondes**

For the meteo data, I used the pre-processor Meteorolagical Data. I converted the data in epw to txt with the tool. 
But I don't understand why does the data change a lot ? (from epw to txt). For exemple it's not the same year.
Where Can I send my zip folder to check what is wrong ?
I have the last version of UMEP an UMEP processing
Thank you
Regards
